src/Uno.UI/UI/Xaml/ResourceDictionary.cs

```diff
@@ -48,7 +72,33 @@ public Uri Source
 		}
 
 		public IList<ResourceDictionary> MergedDictionaries => _mergedDictionaries;
-		public IDictionary<object, object> ThemeDictionaries => _themeDictionaries ??= new ResourceDictionary();
+		public IDictionary<object, object> ThemeDictionaries => GetOrCreateThemeDictionaries();
+
+		private ResourceDictionary GetOrCreateThemeDictionaries()
+		{
+			if (_themeDictionaries is null)
+			{
+				_themeDictionaries = new ResourceDictionary();
+				_themeDictionaries.ResourceDictionaryValueChange += (sender, e) =>
+				{
+					// The purpose of this to updated the cached value when it's no longer valid.
+					// See GetActiveThemeDictionary.
+					if (e.ChangedKey is ResourceKey changedKey && changedKey.Equals(_activeTheme))
+					{
+						// Either a key is added (e.NewValue will be non-null), or
+						// a key is removed (e.NewValue will be null).
+						_activeThemeDictionary = e.NewValue;
```

Comment:
I wonder what happens if the dictionary was falling back on the "Default" theme (or now should be), though we may be getting into the edge case of an edge case.

I also wonder how the case of nested dictionaries in ThemeDictionaries is handled (both by this and by the existing caching logic), since ResourceDictionary.TryGetValue (which is called by GetActiveThemeDictionary()) will search into nested (merged and theme) dictionaries.

I'm curious what the actual scenario was that you ran into where you noticed this bug?

Reply:
> I'm curious what the actual scenario was that you ran into where you noticed this bug?

@davidjohnoliver It was a simple port of NumberBox from microsoft-ui-xaml repo that revealed the bug.

---

> I wonder what happens if the dictionary was falling back on the "Default" theme (or now should be), though we may be getting into the edge case of an edge case.

This is really very interesting. I'll see if I can reproduce in a unit test and look into how would I fix it.

Reply:
I get that it was related to the earlier NumberBox PR, but I was wondering about what was actually happening that triggered this case.

Reply:
@davidjohnoliver I couldn't really know the answer. @jeromelaban Do you know?

Reply:
This is happening because some resources initialized in the NumberBox updates are changing the initialization order, causing some early evaluations. It could be something related to the lazy evaluation you've been looking into @davidjohnoliver.